### ISSUE TITLE
Name translation improvements

### ISSUE THREAD

**Issue by msbarry** [onthegomap/planetiler#86]
[LanguageUtils](https://github.com/onthegomap/planetiler/blob/main/planetiler-basemap/src/main/java/com/onthegomap/planetiler/basemap/util/LanguageUtils.java) is a straight port from openmaptiles logic, but there are a few issues with it.  Please add a comment with any suggestions for improving the logic to assign element names, picking latin/nonlatin names, transliterating, etc. along with some example test cases to illustrate the desired behavior.

**Comment by 1ec5** [comment]
When the name contains no “Latin” characters, the code transliterates the text into Latin text using ICU’s `Any-Latin` transliterator:

https://github.com/onthegomap/planetiler/blob/0d727c4302aa0674df9ca1fa49cb988f84c0f789/planetiler-core/src/main/java/com/onthegomap/planetiler/util/Translations.java#L122-L133

This transliterator can be rough coming from some languages, because different schemes are often used depending on the source language, region, and use case. For example:

[<img src="https://user-images.githubusercontent.com/1231218/155564286-63eeb805-03f9-421f-a87e-18444814f8d9.png" width="400" alt="Kiev">](https://zelonewolf.github.io/openstreetmap-americana/#14.16/50.46092/30.51636)<br>_[Cyrillic text](https://unicode-org.github.io/icu/userguide/transforms/general/#cyrillic) is transliterated to Latin according to the ISO&nbsp;9 standard, which is less biased toward a particular language but still differs from the more common transliteration schemes used in each language or country, as seen here in these [Ukrainian](https://en.wikipedia.org/wiki/Romanization_of_Ukrainian#Conventional_romanization_of_proper_names) park names._

[<img src="https://user-images.githubusercontent.com/1231218/155560968-8d8ed30d-b832-4415-be82-699750b490cf.png" width="400" alt="Shei-Pa">](https://zelonewolf.github.io/openstreetmap-americana/#9.71/24.3934/121.2146)<br>_Chinese text appears to be transliterated to Latin using Hanyu Pinyin. A map intended for lay readers would remove diacritics and spaces between syllables of compound words from these pinyin transliterations. Additionally, in Taiwan, the use of Hanyu Pinyin versus Tongyong Pinyin is [a partisan and regional matter](https://en.wikipedia.org/wiki/Pinyin#In_Taiwan). (These names all end in words like “Ecological Protection Area” that would ideally be translated rather than transliterated.)_

ICU has somewhat more reliable transliterators that require you to know the source language. As in https://github.com/onthegomap/planetiler/issues/14#issuecomment-1041346171, I think knowing the country the feature is in would be a good first step to improving the quality of these transliterations.

**Comment by msbarry** [comment]
Thanks for the incredible detail and pointers @1ec5! The usual use-case I see for `name:latin` and `name:nonlatin` is to provide dual labels when the local name is nonlatin, for example check out the style demos on https://stadiamaps.com/ , like:

<img width="383" alt="image" src="https://user-images.githubusercontent.com/1480504/155642800-3084936a-f58e-4062-b232-b4ab8159f409.png">

On translation/transliteration, I think the preferred solution is for OSM elements to have a latin translation (name:en, name:de, int_name, etc...) - in that case, we won't attempt to transliterate at all. It's just in the case where no latin variation of the name exists that we need to infer it somehow.

**Comment by 1ec5** [comment]
Thanks, that makes sense. The GL style specification even supports [rich text labels](https://maplibre.org/maplibre-gl-js-docs/example/display-and-style-rich-text-labels/), so the second line of these bilingual labels can be formatted differently. Unfortunately, explicit translations are much less likely (and not universally accepted in OSM) for more obscure but common features like street names, so there’s still a high potential for `name:latin` to show up even if `name:en`, `int_name`, etc. are preferred over it.

**Comment by 1ec5** [comment]
As reported [in OSMUS Slack](https://osmus.slack.com/archives/C01V02K52UX/p1645817278260639), Wikidata is making an amusing cameo appearance in some places:

![QIDs](https://user-images.githubusercontent.com/1231218/155787779-d8df31b9-8bd2-4c7b-8007-83bdd34b4e08.png)

[The ways in question](https://www.openstreetmap.org/way/28182832) are exquisitely typeset with en dashes, which the Latin detection code apparently regards as non-Latin, so it falls back to whichever `name:*` it can find that contains only Latin characters. It just happens that these ways are also tagged with `name:etymology:wikidata`, which is guaranteed to be set to an ASCII-only value.

While it may seem contrived to put proper typographical characters on street names, the lack of support for them combined with overeager use of etymology keys can affect other features as well. For example, [this POI in Germany](https://www.openstreetmap.org/node/244987910) is named with [German quotation marks](https://en.wikipedia.org/wiki/Quotation_mark#German), lacks `name:en` or `name:de`, and could conceivably be tagged with `name:etymology:wikidata=Q217964`.

Fortunately, the same regular expression syntax in https://github.com/onthegomap/planetiler/issues/86#issuecomment-1049998723 can avoid these mishaps with some additional [character classes](https://docs.oracle.com/en/java/javase/16/docs/api/java.base/java/util/regex/Pattern.html#ucc): `^[\p{IsLatin}\p{IsPunctuation}]+$`. A lot can be done by combining character classes. If the goal is merely to filter out linguistic content that’s in a different writing system, filtering on the [general category](https://en.wikipedia.org/wiki/Unicode_character_property#General_Category) and script should do the trick:

```regex
^[\P{IsLetter}[\p{IsLetter}&&\p{IsLatin}]]+$
```

This matches anything that isn’t a letter (ideograph, etc.), as well as anything that is a letter in the Latin script.

**Comment by 1ec5** [comment]
Yes, sorry, I thought you had intended this to be an omnibus ticket about localization issues. It would be cleaner to track them in separate issues in the future.

The most robust fix would be to limit the subkeys to those that would be valid BCP 47 codes, like xx and xx-YY and xx-YY-ZZZZ. But limiting it to the languages in the profile should work too.

I think the revised non-Latin detection code would still be worth pursuing regardless. With just the smaller fix you’re suggesting, there will be cases where an alternative language’s name is arbitrarily chosen just because of a “non-Latin” character.

**Comment by msbarry** [comment]
No worries! I did intend this to be an omnibus ticket, just trying to see if I can extract isolated issues from it to work on. I'll give those a shot.

**Comment by gebner** [comment]
> Chinese text appears to be transliterated to Latin using Hanyu Pinyin. A map intended for lay readers would remove diacritics and spaces between syllables of compound words from these pinyin transliterations. Additionally, in Taiwan, the use of Hanyu Pinyin versus Tongyong Pinyin is [a partisan and regional matter](https://en.wikipedia.org/wiki/Pinyin#In_Taiwan). (These names all end in words like “Ecological Protection Area” that would ideally be translated rather than transliterated.)

For Japanese names, the situation is even worse since the Pinyin romanization is not just controversial but incomprehensible.  The following should have been transliterated as Mukogasaki Koen (or Mukogasaki Park, 向ヶ崎公園):

![mukogasakikouen](https://user-images.githubusercontent.com/313929/173233664-927cdcc6-c9c3-4ea9-a69e-2dd987de499a.png)

Romanizing Japanese is nontrivial though, and ICU doesn't support it AFAICT.  We'd need to use a morphological analyzer, like [kuromoji](https://github.com/atilika/kuromoji) (which is far from perfect and would give "kōgasakikōen" in this example).

**Comment by jleedev** [comment]
There are also useless values in `name_de` and `name_en` for some reason.

![Screenshot_20220612-095135](https://user-images.githubusercontent.com/23022/173236614-b971f5e3-1d01-4aa7-beb9-34e724f126fa.png)

**Comment by 1ec5** [comment]
> On translation/transliteration, I think the preferred solution is for OSM elements to have a latin translation (name:en, name:de, int_name, etc...) - in that case, we won't attempt to transliterate at all. It's just in the case where no latin variation of the name exists that we need to infer it somehow.

For many languages, there are keys such as `name:ko-Latn` and `name:sr-Latn` that allow mappers to choose the transliteration system most appropriate to a given language. Ideally the tiles would include those language-qualified property names, because `name:latin` is just as ambiguous as `name`.

Wikidata also has [a variety of properties](https://www.wikidata.org/wiki/Property:P2440#P1659) to indicate the transliteration of a place name, though the long-term approach would be to look at the transliterations stored in [lexicographical data](https://www.wikidata.org/wiki/Wikidata:Lexicographical_data).

**Comment by j9d3it** [comment]
Hi, just finding this issue. For place names in Japanese I've found this Python converter to be really great. https://github.com/polm/cutlet

**Comment by j9d3it** [comment]
Since this issue has come up again recently, I think transliteration should be turned off by default and only enabled with a flag. Or maybe, transliteration should be removed altogether.
